### ISSUE TITLE
DataShard: support for cheap strict-serializable transactions

### ISSUE THREAD

**Issue by snaury** [ydb-platform/ydb#3009]
Our clients are asking for strict serializable isolation, however there's currently no cheap way to implement them. Our coordinators already assign global order to distributed transactions, but it's not cheap: running all writes as distributed transactions, and all reads with global snapshot version, may severely limit throughput (especially for single-threaded clients). What we want, is something that forces all transactions to respect realtime order between all client transactions, yet something that wouldn't limit throughput to plan resolution when storage latency is very small.

In practice, when applying strict serializable to our global transaction order, it boils down to this:

* No write-read reordering. New read transaction must observe all side-effects that have been committed (replied to clients) before the transaction started, including all side-effects that happened before in the global transaction order, anywhere else in the same database. In other words we must use read timestamp that includes any previously (directly or indirectly) committed writes. Read timestamp we currently use for global snapshots is already such a timestamp.
* No write-write reordering. New write-only transaction must happen after all previously (directly or indirectly) committed writes, anywhere else in the same database. This means write tx commit timestamp must be not less than any other commit timestamp we (potentially) replied to clients before transaction started.
* No read-read reodering. When we perform a read in tx1 (abandoning that tx afterwards, i.e. not performing any new reads) and then read in a new tx2, then tx2 must have read timestamp that is not less than that of tx1. This is carefully worded to support advancing read timestamp over the course of the transaction (and when transaction is abandoned there's nothing to advance), although we don't do that, and this is only maybe relevant to read-write transactions where would could potentially side-step some conflicts as long as other locks agree.
* This does not imply any direct order between transactions that intersect in time. For example, we may start a read in tx1 and pause, concurrently commit writes in tx2 (started after the read in tx1), then another read in tx1 might observe changes from tx2, as long as they are consistent. This may be "explained" as if the read timestamp in tx1 advanced past that of tx2, even if in reality something else happens.

When we take global snapshot (read version), we currently query coordinator for the "most recently sent" step to mediators. Shards could not have possibly replied to writes with a step above that, and so read version at the end of that step includes all writes that could have possibly committed anywhere in the same database. We only request this read version when we need a global snapshot, and only use it when we need repeatable reads. Shards wait until their mediator time matches, and all preceding distributed transactions are added to the pipeline, then protect that version from newer writes and perform the read.

Currently when performing "immediate" reads we don't specify any version, and let the shard choose a read version that includes all committed versions at that shard (we call this a "HEAD" read). Any particular shard's view on pending distributed transactions may lag however, so this version may turn out to be lower than that of neighbors. Nothing stops us from always requesting a global read version however (this is "cheap" since it doesn't require waiting for storage as long as coordinators have a valid lease, and supports high throughput due to batching), using that in non-repeatable read mode (immediate reads are usually short, and never need to be repeated). This way we wouldn't block new writes from overwriting the same version, but prevent read-read reordering, and ensure global write-read ordering, i.e. even when shard's mediator time is lagging, it would wait until it receives (and maybe executes, depending on conflicts) all transactions before the given read.

The same goes for immediate writes, datashard currently places them at the most up to date version it knows, but that version may be lagging. This means tx1 may choose v1 (at shard1), while later tx2 may choose v2 (at shard2), where v2 < v1. A concurrent read may choose v3, where v2 < v3 < v1, and observe tx2 happening before tx1. To prevent this we should request a read version as above, but use it as the minimum version for the immediate write. Again, lagging shards will wait until its mediator time catches up, and will perform that write either at the given version or higher (when that version is protected, or its mediator time advanced further).

Important notes:

* Since we have multiple coordinators read timestamp may jump around temporarily, however when the actual read or write is performed is it based on the merged mediator time, and any new read timestamp afterwards will not be less than the given timestamp.
* We may layer multiple reads and writes all to the same version, which is actually desirable (likely no need to wait for future mediator time), and those versions are globally indistinguishable anyway, but provide the most up to date value at any given shard. Such layering is important for maximum single-client throughput, e.g. when performing INSERTs, which consist of a read immediately followed by a write. Multi-shard or repeatable reads will freeze a particular order and bump new immediate writes to higher mediator time.
* Consider we have tx1 commiting value1 at shard1 followed by tx2 commiting value2 at shard2, with a concurrent tx3 trying to read both values. It shouldn't be possible for tx3 to observe value2, but not value1, even though version freezing doesn't happen atomically between multiple shards. When tx3 first observes value2, it means value1 was committed with a version not greater than that of value2, and will be included in followup reads. When tx3 first observes no value1 it means value1 will be bumped to a higher version, and replied when mediator time matches that higher version, so value2 will have version not less than that of value1, and will also not be observable by tx3.

In conclusion what we need is to introduce the notion of "minimum version" for immediate transactions. For reads that would be a non-repeatable read version (we have no reason to choose any newer read version). For writes we would wait until new write mvcc version matches or surpasses that minimum version. This would allow us to have a "strict serializable" feature toggle, which would simply request read version for all transaction, and specify it as a minimum version when performing non-snapshot reads.
